controllers/managedclusterForCGU_controller.go

```diff
@@ -239,12 +266,20 @@ func (r *ManagedClusterForCguReconciler) SetupWithManager(mgr ctrl.Manager) erro
 	return ctrl.NewControllerManagedBy(mgr).
 		Named("managedclusterForCGU").
 		For(&clusterv1.ManagedCluster{},
-			// watch for create and delete events for managedcluster
+			// watch for create event for managedcluster
 			builder.WithPredicates(predicate.Funcs{
 				GenericFunc: func(e event.GenericEvent) bool { return false },
 				CreateFunc:  func(e event.CreateEvent) bool { return true },
 				DeleteFunc:  func(e event.DeleteEvent) bool { return false },
 				UpdateFunc:  func(e event.UpdateEvent) bool { return false },
 			})).
+		Owns(&ranv1alpha1.ClusterGroupUpgrade{},
+			// watch for delete event for owned ClusterGroupUpgrade
+			builder.WithPredicates(predicate.Funcs{
+				GenericFunc: func(e event.GenericEvent) bool { return false },
+				CreateFunc:  func(e event.CreateEvent) bool { return false },
+				DeleteFunc:  func(e event.DeleteEvent) bool { return true },
```

Comment:
With monitoring for the delete is the behavior that we will be enforcing that a CGU will be created for every ManagedCluster until that ManagedCluster has the ztp-done label attached? If so we need to clearly document that for the user. I can see a user getting confused when the delete a CGU and it immediately comes back. 
I would also update the info message on line 248 to read "Found ManagedCluster <managedClusterName> without ztp-done label. Created clusterGroupUpgrade <name>"

Reply:
In general I think it is better to have the user take a direct action that says "I want to restart this". Having a restart be triggered by deleting a CR is a bit non-intuitive. However, including this watch on the delete does have the benefit that the behavior of the controller when restarted matches the running behavior. For example, if a CGU get's deleted this new logic will re-create the CGU. The same would happen if the controller restarted for some reason. Both behaviors are now aligned (which is good). So, I'm ok with leaving it but I think we should add a story to allow the user to affirmatively say that they want to restart the CGU (eg disable and then re-enable the CGU).